sycl/source/exception.cpp

```diff
@@ -10,9 +10,105 @@
 #include <CL/sycl/context.hpp>
 #include <CL/sycl/exception.hpp>
 
+#include <cstring>
+
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 
+namespace { // anonymous
+char reserved_for_errorcode[] =
+    "0SYCL1234567812345678"; // 17 (string terminator plus 'SYCL' marker plus
+                             // error code)
+std::error_code sycl121_proxy_errorcode = make_error_code(sycl::errc::invalid);
+} // namespace
+
+exception::exception(std::error_code ec, const char *Msg)
+    : exception(static_cast<context *>(nullptr), ec, Msg) {}
+
+exception::exception(std::error_code ec, const std::string &Msg)
+    : exception(static_cast<context *>(nullptr), ec, Msg) {}
+
+// new SYCL2020 constructors
+exception::exception(std::error_code ec)
+    : exception(static_cast<context *>(nullptr), ec, "") {}
+
+exception::exception(int ev, const std::error_category &ecat,
+                     const std::string &what_arg)
+    : exception(static_cast<context *>(nullptr), {ev, ecat}, what_arg) {}
+
+exception::exception(int ev, const std::error_category &ecat,
+                     const char *what_arg)
+    : exception(static_cast<context *>(nullptr), {ev, ecat},
+                std::string(what_arg)) {}
+
+exception::exception(int ev, const std::error_category &ecat)
+    : exception(static_cast<context *>(nullptr), {ev, ecat}, "") {}
+
+exception::exception(context ctx, std::error_code ec,
+                     const std::string &what_arg)
+    : exception(&ctx, ec, what_arg) {}
+
+exception::exception(context ctx, std::error_code ec, const char *what_arg)
+    : exception(ctx, ec, std::string(what_arg)) {}
+
+exception::exception(context ctx, std::error_code ec)
+    : exception(ctx, ec, "") {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat,
+                     const char *what_arg)
+    : exception(ctx, {ev, ecat}, std::string(what_arg)) {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat,
+                     const std::string &what_arg)
+    : exception(ctx, {ev, ecat}, what_arg) {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat)
+    : exception(ctx, ev, ecat, "") {}
+
+// protected base constructor for all SYCL2020 constructors
+exception::exception(context *ctxPtr, std::error_code ec,
+                     const std::string &what_arg)
+    : MMsg(what_arg + reserved_for_errorcode), MCLErr(PI_INVALID_VALUE),
+      MContext(ctxPtr) {
+  // For compatibility with previous implementation, we are "hiding" the
+  // std:::error_code in the MMsg string, behind the null string terminator
+  size_t whatLen = what_arg.length();
+  char *reservedPtr = &MMsg[whatLen];
+  reservedPtr[0] = '\0';
+  reservedPtr++;
+  // insert 'SYCL' marker.
+  reservedPtr[0] = 'S';
+  reservedPtr[1] = 'Y';
+  reservedPtr[2] = 'C';
+  reservedPtr[3] = 'L';
+  reservedPtr += 4;
```

Comment:
Could you please clarify why the marker is needed? Isn't additional `\0` enough?
We could construct the following string:
`Internal error! 128 RAM is not enough for creating queue object!\0[error_code value]\0`
And in order to find "reserved area" we just need to find the second `\0` starting from the end the string.

Reply:
As I noted in other comment this idea doesn't fly if exception is constructed with `exception(int, const std::error_category &, const std::string &);`.

Reply:
@romanovvlad  I put in the marker in the event we are passed a SYCL 1.2.1 exception, but I'll see if I can drop it.

Reply:
@bader  -  I have tests where all 12 of the new constructors are called, and I call them initially with empty strings just to make sure the string handling is ok.  I reworked that bit a lot though, based on your other feedback.  Let me know if I'm not understanding.    basic_tests/exceptions-SYCL-2020.cpp

Reply:
I've suggested a new test case to demonstrate the issue.

Reply:
added the new case. thanks.